### PR TITLE
 osd/scrub: modify "scrub" & "deep_scrub" ASOK commands

### PR DESCRIPTION
**Pull request by ronen-fr**
Fix all instances of 'deep_scrub' to the documented format ('deep-scrub');
    
Make all formats of the documented command ('ceph tell $pg [deep-]scrub', and the old 'ceph pg scrub $pgid') trigger an operator-priority scrub;
    
Introduce a dev/tests-only format for the required functionality of "trigger a scrub by modifying the timestamps".
    
Change some implementation details.

### CODE REVIEW

**Comment by athanatos** [comment]
I've always interpreted the scrub and deep_scrub asok commands to convey an intent to force a scrub.  IIRC, messing with the last-scrub stamp was just a super lazy way of forcing the PG to do a scrub quickly -- it wasn't a behavior the user is supposed to actually depend on.  Do you have a link to the documentation that describes the "time offset"?

**Comment by ronen-fr** [comment]
@athanatos :
> Do you have a link to the documentation that describes the "time offset"?
Looking for it (was already searching for it, as it has to be changed to mention the 'force').

File: src/osd/OSD.cc [hunk 1]
```diff
@@ -4363,18 +4363,20 @@ void OSD::final_init()
...
   r = admin_socket->register_command(
-    "pg "			   \
-    "name=pgid,type=CephPgid "	   \
-    "name=cmd,type=CephChoices,strings=scrub " \
-    "name=time,type=CephInt,req=false",
+    "pg "
+    "name=pgid,type=CephPgid "
+    "name=cmd,type=CephChoices,strings=scrub "
+    "name=time,type=CephInt,req=false "
+    "name=force,type=CephBool,req=false",
```
Comment (rzarzynski) [inline src/osd/OSD.cc:4370]: ACK. So we're just adding the `force` param. Other lines are changes for the sake of backslash cleaning.

File: src/osd/OSD.cc [hunk 2]
```diff
@@ -4363,18 +4363,20 @@ void OSD::final_init()
...
   r = admin_socket->register_command(
-    "pg "			   \
-    "name=pgid,type=CephPgid "	   \
-    "name=cmd,type=CephChoices,strings=deep_scrub " \
-    "name=time,type=CephInt,req=false",
+    "pg "
+    "name=pgid,type=CephPgid "
+    "name=cmd,type=CephChoices,strings=deep_scrub "
+    "name=time,type=CephInt,req=false "
+    "name=force,type=CephBool,req=false",
```
Comment (rzarzynski) [inline src/osd/OSD.cc:4379]: ditto.

File: src/osd/PrimaryLogPG.cc [hunk 3]
```diff
@@ -1158,39 +1158,19 @@ void PrimaryLogPG::do_command(
...
-      // Instead of marking must_scrub force a schedule scrub
-      utime_t stamp = ceph_clock_now();
-      if (time == 0)
-        stamp -= scrub_max_interval;
-      else
-        stamp -=  (float)time;
-      stamp -= 100.0;  // push back last scrub more for good measure
-      if (deep) {
-        set_last_deep_scrub_stamp(stamp);
+        m_scrubber->on_operator_periodic_cmd(f.get(), deep, time);
```
Comment (rzarzynski) [inline src/osd/PrimaryLogPG.cc:1172]: OK, now it dispatches.

File: src/osd/scrubber/pg_scrubber.cc [hunk 4]
```diff
@@ -656,6 +656,77 @@ Scrub::sched_conf_t PgScrubber::populate_config_params() const
...
+void asok_response_section(
+    ceph::Formatter* f,
+    bool is_periodic,
+    scrub_level_t scrub_level
+    /*const char* section_value*/)
+{
+  f->open_object_section("result");
+  f->dump_bool("deep", (scrub_level == scrub_level_t::deep));
+  f->dump_bool("must", !is_periodic);
+  f->close_section();
```
Comment (rzarzynski) [inline src/osd/scrubber/pg_scrubber.cc:671]: We stopped dumpig the `stamp` field though the information in the period variant is still available. Is this intentional?

**Review by rzarzynski — COMMENTED** [review]
At first glance looks good. A question about the dumping format change.

File: src/osd/scrubber/pg_scrubber.cc (hunk 4 above)
Comment (ronen-fr) [reply src/osd/scrubber/pg_scrubber.cc:671]: @rzarzynski - I am redoing this whole thing:

- the scrub and deep-scrub commands will have no parameters, and will always force a scrub

- the test/debug behavior, that is used extensively by tests and during debug, will have its own command.

No need to continue reviewing this version.

Comment (ronen-fr) [reply src/osd/scrubber/pg_scrubber.cc:670]: Re the "stamp" - a good point. I have removed it as it only exists in one situation.
Will restore, just for the case that there was someone using that.

Comment (rzarzynski) [reply src/osd/scrubber/pg_scrubber.cc:672]: Sure, please let me know after it's ready for another round.

File: src/osd/OSD.cc (hunk 1 above)
Comment (ronen-fr) [reply src/osd/OSD.cc]: indeed so.

File: src/osd/scrubber/pg_scrubber.cc (hunk 4 above)
Comment (ronen-fr) [reply src/osd/scrubber/pg_scrubber.cc:672]: @rzarzynski  - it is now, I hope. The comments are from before I changed the PR into a draft, modified it, run some Teu tests, and now marked it Ready again.

Comment (ronen-fr) [reply src/osd/scrubber/pg_scrubber.cc:672]: (and I have restored the 'stamp' field, as you've suggested)

**Comment by ronen-fr** [comment]
I will be merging following multiple Teuthology tests
(as wip-rf-force-scrubs & rf-repl-hp-onto-asok-force),
and as the PR mostly affects test-related commands.

File: src/osd/scrubber/scrub_job.h [hunk 5]
```diff
@@ -178,4 +231,19 @@ struct formatter<Scrub::ScrubJob> {
...
   }
 };
+
+template <>
+struct formatter<Scrub::sched_conf_t> {
+  constexpr auto parse(format_parse_context& ctx) { return ctx.begin(); }
+  template <typename FormatContext>
+  auto format(const Scrub::sched_conf_t& cf, FormatContext& ctx)
+  {
+    return format_to(
```
Comment (cbodley) [inline src/osd/scrubber/scrub_job.h:241]: i've opened https://github.com/ceph/ceph/pull/54529 to fix a ftbfs here on gcc13

**Pull request merged**